src/mscorlib/corefx/System/Buffers/DefaultArrayPool.cs

```diff
@@ -125,37 +153,133 @@ public override void Return(T[] array, bool clearArray = false)
             {
                 throw new ArgumentNullException(nameof(array));
             }
-            else if (array.Length == 0)
-            {
-                // Ignore empty arrays.  When a zero-length array is rented, we return a singleton
-                // rather than actually taking a buffer out of the lowest bucket.
-                return;
-            }
 
             // Determine with what bucket this array length is associated
             int bucket = Utilities.SelectBucketIndex(array.Length);
 
-            // If we can tell that the buffer was allocated, drop it. Otherwise, check if we have space in the pool
+            // If we can tell that the buffer was allocated (or empty), drop it. Otherwise, check if we have space in the pool.
             if (bucket < _buckets.Length)
             {
-                // Clear the array if the user requests
+                // Clear the array if the user requests.
                 if (clearArray)
                 {
                     Array.Clear(array, 0, array.Length);
                 }
 
-                // Return the buffer to its bucket.  In the future, we might consider having Return return false
-                // instead of dropping a bucket, in which case we could try to return to a lower-sized bucket,
-                // just as how in Rent we allow renting from a higher-sized bucket.
-                _buckets[bucket].Return(array);
+                // Check to see if the buffer is the correct size for this bucket
+                GlobalBucket b = _buckets[bucket];
+                if (array.Length < b.BufferLength)
+                {
+                    throw new ArgumentException(SR.ArgumentException_BufferNotFromPool, nameof(array));
```

Comment:
> For an empty array bucket would be 0

Actually, no, because of the selection algorithm and the -1 used in https://github.com/dotnet/coreclr/blob/master/src/mscorlib/corefx/System/Buffers/Utilities.cs#L17, 0 ends up looking like a gigantic buffer and skips past this whole block.